gel/auth/email_password.py

```diff
@@ -0,0 +1,550 @@
+                identity_id=token_data.identity_id,
+            )
+        else:
+            logger.info(
+                "no code in register response, "
+                "assuming verification required"
+            )
+            logger.debug("PKCE verifier: %s", pkce.verifier)
+            return SignUpVerificationRequiredResponse(
+                verifier=pkce.verifier,
+                identity_id=register_json.get("identity_id"),
+            )
+
+    async def _sign_in(self, email: str, password: str) -> SignInResponse:
+        logger.info("signing in user: %s", email)
+        pkce = self._generate_pkce()
+        sign_in_response = await self._http_request(
+            "POST",
+            "/authenticate",
+            json={
+                "email": email,
+                "provider": self.provider,
+                "password": password,
+                "challenge": pkce.challenge,
+            },
+        )
+        try:
+            sign_in_response.raise_for_status()
+        except httpx.HTTPStatusError as e:
+            logger.error("sign in error: %s", e)
+            return SignInFailedResponse(
+                verifier=pkce.verifier,
+                status_code=e.response.status_code,
+                message=e.response.text,
+            )
+        sign_in_json = sign_in_response.json()
+        if "error" in sign_in_json:
+            error = sign_in_json["error"]
+            logger.error("sign in error: %s", error)
+            return SignInFailedResponse(
+                verifier=pkce.verifier,
+                status_code=sign_in_response.status_code,
+                message=error,
+            )
+        elif "code" in sign_in_json:
+            code = sign_in_json["code"]
+            logger.info("exchanging code for token: %s", code)
+            token_data = await pkce.internal_exchange_code_for_token(code)
+
+            logger.debug("PKCE verifier: %s", pkce.verifier)
+            logger.debug("token data: %s", token_data)
+            return SignInCompleteResponse(
+                verifier=pkce.verifier,
+                token_data=token_data,
+                identity_id=token_data.identity_id,
+            )
+        else:
+            logger.info(
+                "no code in sign in response, assuming verification required"
+            )
+            logger.debug("PKCE verifier: %s", pkce.verifier)
+            return SignInVerificationRequiredResponse(
+                verifier=pkce.verifier,
+                identity_id=sign_in_json.get("identity_id"),
+            )
+
+    async def _verify_email(
+        self, verification_token: str, verifier: Optional[str]
+    ) -> EmailVerificationResponse:
+        logger.info("verifying email")
+        verify_response = await self._http_request(
+            "POST",
+            "/verify",
+            json={
+                "verification_token": verification_token,
+                "provider": self.provider,
+            },
+        )
```

Comment:
Is it expected that the verification will proceed (even successfully) with or without a valid PKCE verifier?

Reply:
Yes, that's right. PKCE just protects the actual auth token exchange, so you can verify an email from any user agent, but you will not be able to then parlay that into an authentication unless you have the PKCE verifier.